### PR TITLE
Add sequence Presto function

### PR DESCRIPTION
**Pull request by kewang1024**
Added 
- [x] sequence(bigint, bigint) -> array(bigint)
- [x] sequence(bigint, bigint, bigint) -> array(bigint)

Since this is my first velox function, I will first try to finish this functionality alone in this PR before it gets too big and the rest can be a followup
- [ ] [Optimization] Add fast path when step is constant
- [ ] sequence(date, date)
- [ ] sequence(date, date, interval day to second)
- [ ] sequence(date, date, interval year to month)
- [ ] sequence(timestamp, timestamp, interval day to second)
- [ ] sequence(timestamp, timestamp, interval year to month)

### CODE REVIEW

File: velox/functions/prestosql/Sequence.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+} // namespace
+
+VELOX_DECLARE_VECTOR_FUNCTION(
+    udf_sequence,
+    SequenceFunction::signatures(),
+    std::make_unique<SequenceFunction>());
+
+} // namespace facebook::velox::functions
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:94]: Format: needs an empty line in the end.

File: velox/functions/prestosql/Sequence.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    // todo integer, integer, integer -> array(integer)
+    signatures.emplace_back(exec::FunctionSignatureBuilder()
+                                .returnType("array(integer)")
+                                .argumentType("integer")
+                                .argumentType("integer")
+                                .argumentType("integer")
+                                .build());
+
+    // date, date -> array(date)
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:43]: Are we going to add date support in a second PR?

File: velox/functions/prestosql/Sequence.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ private:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto pool = context.pool();
+    auto startVector = args[0]->asFlatVector<int32_t>();
+    auto endVector = args[1]->asFlatVector<int32_t>();
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:56]: We don't use the 3rd parameter (step) yet?

File: velox/functions/prestosql/Sequence.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    context.ensureWritable(rows, outputType, result);
+    result->clearNulls(rows);
+    auto arrayResult = result->as<ArrayVector>();
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:59]: Not sure if right now or later (probably later), it is worth to learn if the most common use case is with some arguments (or all) being constants and have optimized code for such case.

File: velox/functions/prestosql/Sequence.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    auto baseOffset = elementsResult->size();
+    vector_size_t offset = baseOffset;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto start = startVector->valueAt(row);
+      auto end = endVector->valueAt(row);
+      rawOffsets[row] = offset;
+      rawSizes[row] = end - start + 1;
+      elementsResult->resize(offset + rawSizes[row]);
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:75]: It might be beneficial to run the applyToSelected() once to determine the total number of elements and elementsResult->reserve() before we add elements to it.

File: velox/functions/prestosql/Sequence.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      for (auto i = start; i <= end; i++) {
+        auto vector =
+            BaseVector::create<FlatVector<int32_t>>(BIGINT(), 1, pool);
+        auto flatVector = vector->asFlatVector<int32_t>();
+        flatVector->mutableRawValues()[0] = i;
+        elementsResult->copy(flatVector, offset++, row, 1);
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:81]: From what I understand here we create a 1-element vector for every element and then copy it (single item) to the target elements vector.
This would be very inefficient. We should write to elements directly.

**Comment by spershin** [comment]
@kewang1024 Thank you for working on this! :)

File: velox/docs/functions/presto/coverage.rst [hunk 7]
```diff
@@ -239,7 +239,7 @@ Here is a list of all scalar and aggregate Presto functions with functions that
     :func:`array_except`                      :func:`day_of_year`                       json_extract                              :func:`rtrim`                             st_within                                     :func:`bitwise_and_agg`                       :func:`ntile`
     :func:`array_frequency`                   :func:`degrees`                           :func:`json_extract_scalar`               scale_qdigest                             st_x                                          :func:`bitwise_or_agg`                        :func:`percent_rank`
     :func:`array_has_duplicates`              :func:`dow`                               :func:`json_format`                       :func:`second`                            st_xmax                                       :func:`bool_and`                              :func:`rank`
-    :func:`array_intersect`                   :func:`doy`                               :func:`json_parse`                        sequence                                  st_xmin                                       :func:`bool_or`                               :func:`row_number`
```
Comment (mbasmanova) [inline velox/docs/functions/presto/coverage.rst:242]: Updates to coverage maps needs to be done in a separate PR.

File: velox/functions/prestosql/Sequence.cpp [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    VELOX_FAIL("Unsupported Type")
+  }
+}
+
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  static const std::map<std::string, std::string> typePairs = {
+      {"integer", "integer"}, {"bigint", "bigint"}};
+  std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+  signatures.reserve(typePairs.size());
+  for (const auto& [argType, arg2Type] : typePairs) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:237]: Looks like types of 2 arguments are always match. It might be simpler to write this as

```
for (const auto& type : {"integer", "bigint"}) {
   signatures.emplace_back(exec::FunctionSignatureBuilder()
                                .returnType("array(bigint)")
                                .argumentType(type)
                                .argumentType(type)
                                .build());
}
```

File: velox/functions/prestosql/Sequence.cpp [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+#include <iostream>
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/ConstantVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+template <typename TInput1, typename TInput2>
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:25]: Looks like TInput1 and TInput2 are always the same. Let's change to have just one template parameter T.

File: velox/functions/prestosql/Sequence.cpp [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+  }
+
+  VectorPtr applyConstant(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    auto startVector = args[0]->as<ConstantVector<int64_t>>();
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:58]: This won't work if input types are INTEGER. Do we have a test case for that?

**Review by mbasmanova — COMMENTED** [review]
@kewang1024 Some initial comments.

File: velox/functions/prestosql/tests/SequenceTest.cpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  testExpression("sequence(C0, C1)", {startVector, endVector}, expected);
+}
+
+TEST_F(SequenceTest, testConstant) {
+  const auto endVector = makeNullableFlatVector<int>({2, 5});
+  VectorPtr expected = makeArrayVector<long long>({{1, 2}, {1, 2, 3, 4, 5}});
+  testExpression("sequence(1, C0)", {endVector}, expected);
+}
+
+TEST_F(SequenceTest, testInvalidResult) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/SequenceTest.cpp:63]: > testInvalidResult

Looks like this test is for invalid input, not result. Consider renaming to "invalidInput".

File: velox/functions/prestosql/tests/SequenceTest.cpp [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  const auto startVector = makeNullableFlatVector<int>({1});
+  const auto endVector = makeNullableFlatVector<int>({100000});
+  testExpressionWithError(
+      "sequence(C0, C1)",
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/SequenceTest.cpp:67]: Maybe just "sequence(1, 100000)"

File: velox/functions/prestosql/tests/SequenceTest.cpp [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+      {startVector, endVector},
+      "etConnectorFactory");
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/SequenceTest.cpp:69]: This error message is rather strange. It would be nice to ensure that error message is user-friendly.

File: velox/functions/prestosql/tests/SequenceTest.cpp [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      const std::vector<VectorPtr>& input,
+      const std::string& expectedError) {
+    VELOX_ASSERT_THROW(
+        evaluate(expression, makeRowVector(input)), expectedError);
+  }
+};
+} // namespace
+
+TEST_F(SequenceTest, sequence) {
+  const auto startVector = makeNullableFlatVector<int>({1, 2, 10});
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/SequenceTest.cpp:47]: - Use makeFlatVector when there are no nulls.
- int -> int64_t or int32_t

File: velox/functions/prestosql/tests/SequenceTest.cpp [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+      const VectorPtr& expected) {
+    auto result = evaluate(expression, makeRowVector(input));
+    assertEqualVectors(expected, result);
+  }
+
+  void testExpressionWithError(
+      const std::string& expression,
+      const std::vector<VectorPtr>& input,
+      const std::string& expectedError) {
+    VELOX_ASSERT_THROW(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/SequenceTest.cpp:40]: When testing failing expressions, please, make sure to also test that adding TRY makes failures go away and produces "correct" results or nulls.

File: velox/functions/prestosql/Sequence.cpp (hunk 9 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:25]: As of now, they're the same, but I assumed in the future, we might need the function to also support other scenarios like sequence(int, long)? 
If that's the case, we should remain this way?

File: velox/functions/prestosql/Sequence.cpp [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+
+    vector_size_t elementOffset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      rawOffsets[row] = elementOffset;
+      rawSizes[row] = elementsCount;
+      writeToElements(rawElements, start, end, elementOffset);
+      elementOffset += elementsCount;
+    });
+
+    BufferPtr nulls;
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:87]: This variable is not needed. Just use nullptr.

File: velox/functions/prestosql/Sequence.cpp [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+    if (startVector->isNullAt(0) || endVector->isNullAt(0)) {
+      // If start or end is a null constant, the result should be all nulls.
+      return BaseVector::createNullConstant(outputType, numRows, pool);
+    }
+
+    int start = startVector->valueAt(0);
+    int end = endVector->valueAt(0);
+    const auto elementsCount = end - start + 1;
+    if (elementsCount <= 0) {
+      return BaseVector::createNullConstant(outputType, numRows, pool);
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:69]: Is this the right semantic? Here is what I see in Presto:

```
presto> select sequence(10, 5);
        _col0
---------------------
 [10, 9, 8, 7, 6, 5]

presto> select sequence(5, 10);
        _col0
---------------------
 [5, 6, 7, 8, 9, 10]

```

File: velox/functions/prestosql/Sequence.cpp (hunk 9 above)
Comment (spershin) [reply velox/functions/prestosql/Sequence.cpp:25]: @mbasmanova 
@kewang1024

Indeed, seems start/stop and step can be of different type:

Generate a sequence of dates from start to stop, incrementing by step. The type of step can be either INTERVAL DAY TO SECOND or INTERVAL YEAR TO MONTH.

File: velox/functions/prestosql/Sequence.cpp (hunk 10 above)
Comment (spershin) [reply velox/functions/prestosql/Sequence.cpp:58]: I think @mbasmanova meant "if input types are NOT integer".

File: velox/functions/prestosql/Sequence.cpp (hunk 9 above)
Comment (mbasmanova) [reply velox/functions/prestosql/Sequence.cpp:25]: @kewang1024 I don't see such signatures in Presto. Let's use a single template parameter and change that if/when we need to. I also don't see Presto allowing 'integer' as input parameter. Only 'bigint' seems to be allowed. This suggests that we don't need a template parameter at all.

```
presto> show functions like 'sequence';
 Function |   Return Type    |                Argument Types                | Function Type | Deterministic |                  Description                   | Va>
----------+------------------+----------------------------------------------+---------------+---------------+------------------------------------------------+--->
 sequence | array(bigint)    | bigint, bigint                               | scalar        | true          |                                                | fa>
 sequence | array(bigint)    | bigint, bigint, bigint                       | scalar        | true          | Sequence function to generate synthetic arrays | fa>
 sequence | array(date)      | date, date                                   | scalar        | true          |                                                | fa>
 sequence | array(date)      | date, date, interval day to second           | scalar        | true          |                                                | fa>
 sequence | array(date)      | date, date, interval year to month           | scalar        | true          |                                                | fa>
 sequence | array(timestamp) | timestamp, timestamp, interval day to second | scalar        | true          |                                                | fa>
 sequence | array(timestamp) | timestamp, timestamp, interval year to month | scalar        | true          |                                                | fa>
```

File: velox/functions/prestosql/Sequence.cpp (hunk 17 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:69]: Good catch!

File: velox/functions/prestosql/Sequence.cpp (hunk 9 above)
Comment (mbasmanova) [reply velox/functions/prestosql/Sequence.cpp:25]: @spershin 

> seems start/stop and step can be of different type:

This is the case, but it doesn't look like template the class would be sufficient to support these. I suggest we either remove template parameter or add support for all the signatures.

File: velox/functions/prestosql/Sequence.cpp [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    auto startVector = args[0]->as<ConstantVector<int64_t>>();
+    auto endVector = args[1]->as<ConstantVector<int64_t>>();
+    if (startVector->isNullAt(0) || endVector->isNullAt(0)) {
+      // If start or end is a null constant, the result should be all nulls.
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp]: Can you elaborate where we define default null behavior?

File: velox/functions/prestosql/Sequence.cpp (hunk 10 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp]: In my understanding, it would be guarded by function signatures?

File: velox/functions/prestosql/tests/SequenceTest.cpp (hunk 15 above)
Comment (kewang1024) [reply velox/functions/prestosql/tests/SequenceTest.cpp:40]: I'm not sure I understand this one, this is copied from other functions tests we already have

File: velox/docs/functions/presto/array.rst [hunk 19]
```diff
@@ -214,6 +214,14 @@ Array Functions
     Returns a subarray starting from index ``start``(or starting from the end
     if ``start`` is negative) with a length of ``length``.
 
+.. function:: sequence(start, end) -> array
```
Comment (spershin) [inline velox/docs/functions/presto/array.rst:217]: end -> stop.
Should we just copy the description from Presto documentation, seems like it is more clear there?
https://prestodb.io/docs/current/functions/array.html#id5

File: velox/functions/prestosql/Sequence.cpp [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    if (args[0]->type() == BIGINT()) {
+      localResult = applyNumber(rows, args, outputType, context);
+    }
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:39]: Does this mean that for all other types localResult would be empty?
Should we put some VELOX_FAIL() in the "else {}"?

File: velox/functions/prestosql/Sequence.cpp [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  VectorPtr applyNumber(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    int totalCount = 0;
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:50]: I see the meaning of 'totalCount' is the total number of elements?
Should we name it 'numElements'?

Also, this will be used in BaseVector::create(), which expects "vector_size_t" type, so lets use that instead of int.

File: velox/functions/prestosql/Sequence.cpp [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+    auto startVector = args[0];
+    auto stopVector = args[1];
+    VectorPtr stepVector;
+    if (args.size() == 3) {
+      stepVector = args[2];
+    }
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      int count = sequenceCount(startVector, stopVector, stepVector, row);
+      totalCount += count;
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:61]:
```
totalCount += sequenceCount(startVector, stopVector, stepVector, row);
```
?

File: velox/functions/prestosql/Sequence.cpp [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    int totalCount = 0;
+
+    auto startVector = args[0];
+    auto stopVector = args[1];
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:53]: If these vectors have encodings and below we pass them to sequenceCount(), then we would likely perform expensive computation or decoding per row.
Check how other UDF implementations get decoded vectors, for example Split.cpp:100

File: velox/functions/prestosql/Sequence.cpp [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  int sequenceCount(
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:92]: VectorPtr is:
```
using VectorPtr = std::shared_ptr<BaseVector>;
```

Passing it by value will incur ref increase and then decrease, which would be a performance hit, especiall done per row.
Consider passing by const ref:
```
const VectorPtr& startVector,
...
```

File: velox/functions/prestosql/Sequence.cpp [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+          stepVector,
+          row);
+      elementsOffset += rawSizes[row];
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  int sequenceCount(
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:89]: Maybe sequenceSize()?

File: velox/functions/prestosql/Sequence.cpp [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
+      vector_size_t row) const {
+    long start = valueInVector(startVector, row);
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:94]: When using decoded vectors it would look something like this:
```
startVector->valueAt<int64_t>(row);
```

File: velox/functions/prestosql/Sequence.cpp [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+
+  long valueInVector(VectorPtr baseVector, vector_size_t row) const {
+    if (baseVector->isConstantEncoding()) {
+      return baseVector->as<ConstantVector<long long>>()->valueAt(0);
+    } else {
+      return baseVector->asFlatVector<long long>()->valueAt(row);
+    }
+  }
+
+  void checkArguments(int start, int stop, int step) const {
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:114]: When a class method does not use any class instance's members it is better to mark it with "static", "const" won't be needed in this case:
```
static void checkArguments(int start, int stop, int step) {
```

File: velox/functions/prestosql/Sequence.cpp [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+        "sequence stop value should be greater than or equal to start value if "
+        "step is greater than zero otherwise stop should be less than or equal to start")
+    VELOX_USER_CHECK_LE(
+        (stop - start) / step + 1,
+        kMaxResultEntries,
+        "result of sequence function must not have more than 10000 entries");
+  }
+
+  void writeToElements(
+      long long* elements,
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:127]: We don't use "long long", but "int64_t".

File: velox/functions/prestosql/Sequence.cpp [hunk 29]
```diff
@@ -0,0 +1,173 @@
...
+      int elementsOffset,
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:131]: Same about references and decoded vectors.
But, actually, when using decoded vectors you'll be dealing with pointers, so const ref is not a problem:
```
DecodedVector* starts,
...
```

File: velox/functions/prestosql/Sequence.cpp [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+      step = stop >= start ? 1 : -1;
+    } else {
+      step = valueInVector(stepVector, row);
+    }
+
+    long length = (stop - start) / step + 1;
+    for (long i = 0, value = start; i < length; ++i, value += step) {
+      elements[elementsOffset] = value;
+      elementsOffset++;
+    }
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:146]: Would this be a little bit more clear:
```
const auto numElements = (stop - start) / step + 1;
const auto valueEnd = start + stop * numElements;
for (auto value = start, value != valueEnd; value += step) {
  elements[elementsOffset++] = value;
}
```
?

Also this function is quite simple - it might be better to just leave its body in the context.applyToSelectedNoThrow's lambda.

File: velox/functions/prestosql/tests/SequenceTest.cpp (hunk 15 above)
Comment (spershin) [reply velox/functions/prestosql/tests/SequenceTest.cpp:40]: @kewang1024 
I believe Masha talks about wrapping test expression "sequence(C0, C1)" into TRY and making sure it works properly with TRY on top:

"TRY(sequence(C0, C1))".

Some other tests should have that.

**Review by spershin — COMMENTED** [review]
It is getting close. :)
Sorry, lots of comments, but we've all been there! :)

File: velox/functions/prestosql/Sequence.cpp (hunk 25 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:89]: I'm also pretty torn between those two names, but since we want to imply the count number in the sequence instead of the overall byte size of the sequence, I'm slightly towards `sequenceCount`
Let me know if you have strong opinions towards it

File: velox/functions/prestosql/Sequence.cpp (hunk 30 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:146]: Just changed the logic
But I haven't moved the logic into the lambda, because from the java Sequence function, this function could be used in multiple places (https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/SequenceFunction.java#L147), so I think we would be reuse this function soon when we add more support

File: velox/functions/prestosql/Sequence.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/functions/prestosql/Sequence.cpp]: @kewang1024 Check out documentation at https://facebookincubator.github.io/velox/develop/scalar-functions.html . By default, all functions are assumes to have default null behavior. This can be changed by overriding isDefaultNullBehavior() method.

File: velox/functions/prestosql/tests/SequenceTest.cpp (hunk 15 above)
Comment (mbasmanova) [reply velox/functions/prestosql/tests/SequenceTest.cpp:40]: @spershin That's what I meant. Not all tests are following this best practice, but they should.

File: velox/functions/prestosql/Sequence.cpp [hunk 31]
```diff
@@ -0,0 +1,173 @@
...
+      DecodedVector* stepVector,
+      vector_size_t row) const {
+    auto start = startVector->valueAt<int64_t>(row);
+    auto stop = stopVector->valueAt<int64_t>(row);
+    int64_t step;
+    if (stepVector == nullptr) {
+      step = stop >= start ? 1 : -1;
+    } else {
+      step = stepVector->valueAt<int64_t>(row);
+    }
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:109]: Leaving uninitialized variable can later lead to bugs if code after it is changed.
We can leverage ? operator here:
```
  const int64_t step = (stepVector == nullptr)
      ? (stop >= start ? 1 : -1)
      : step = stepVector->valueAt<int64_t>(row);
```

or 

```
int64_t step{0};
if() else ();
```

File: velox/functions/prestosql/Sequence.cpp [hunk 32]
```diff
@@ -0,0 +1,173 @@
...
+    checkArguments(start, stop, step);
+    return (vector_size_t)((stop - start) / step + 1);
+  }
+
+  static void checkArguments(int start, int stop, int step) {
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:114]: int -> int64_t

File: velox/functions/prestosql/Sequence.cpp [hunk 33]
```diff
@@ -0,0 +1,173 @@
...
+          stepVector,
+          row);
+      elementsOffset += rawSizes[row];
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  vector_size_t sequenceCount(
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:97]: static member function?

File: velox/functions/prestosql/Sequence.cpp [hunk 34]
```diff
@@ -0,0 +1,173 @@
...
+        VELOX_FAIL("Unsupported type for argument of 'sequence' function");
+      }
+    } catch (const std::exception& e) {
+      context.setErrors(rows, std::current_exception());
+      return;
+    }
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+  VectorPtr applyNumber(
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:51]: static member function?

File: velox/functions/prestosql/Sequence.cpp [hunk 35]
```diff
@@ -0,0 +1,173 @@
...
+        step > 0 ? stop >= start : stop <= start,
+        "sequence stop value should be greater than or equal to start value if "
+        "step is greater than zero otherwise stop should be less than or equal to start")
+    VELOX_USER_CHECK_LE(
+        (stop - start) / step + 1,
+        kMaxResultEntries,
+        "result of sequence function must not have more than 10000 entries");
+  }
+
+  void writeToElements(
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:126]: static member function?

**Review by spershin — COMMENTED** [review]
Thank you for the update!
Few more comments!

**Review by spershin — APPROVED** [review]
Looks good! :)
Congrats with the first Velox function!
The next steps would be:
1. Adding integration test for this function in Presto Native.
2. Support other types (date).

File: velox/functions/prestosql/Sequence.cpp [hunk 36]
```diff
@@ -0,0 +1,168 @@
...
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    try {
+      if (args[0]->type() == BIGINT()) {
+        localResult = applyNumber(rows, args, outputType, context);
+      } else {
+        VELOX_NYI(
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:41]: This check is not necessary, since the signature doesn't allow for this. Please, remove.

File: velox/functions/prestosql/Sequence.cpp [hunk 37]
```diff
@@ -0,0 +1,168 @@
...
+            "Unsupported type for argument of 'sequence' function: ",
+            args[0]->type()->toString());
+      }
+    } catch (const std::exception& e) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:45]: What types of errors are we catching here?

File: velox/functions/prestosql/Sequence.cpp [hunk 38]
```diff
@@ -0,0 +1,168 @@
...
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto startVector = decodedArgs.at(0);
+    auto stopVector = decodedArgs.at(1);
+    DecodedVector* stepVector = nullptr;
+    if (args.size() == 3) {
+      stepVector = decodedArgs.at(2);
+    }
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:70]: If some rows generated an error, we need to make sure not to process these in the next loop. See usage of context.deselectErrors in Map.cpp

**Review by mbasmanova — CHANGES_REQUESTED** [review]
@kewang1024 There are still a few things that need to be taken care of.

File: velox/functions/prestosql/Sequence.cpp (hunk 36 above)
Comment (spershin) [reply velox/functions/prestosql/Sequence.cpp:41]: @mbasmanova 
I assume this check if for the future code, when DATE would be supported.
So it is required.
The else with NYI is not required, I agree.

File: velox/functions/prestosql/Sequence.cpp (hunk 37 above)
Comment (spershin) [reply velox/functions/prestosql/Sequence.cpp:45]: I was surprised that we are catching a generic exception.
But the code uses current_exception and it seems that's what we do everywhere.
I believe this is to support "try".

File: velox/functions/prestosql/Sequence.cpp (hunk 38 above)
Comment (spershin) [reply velox/functions/prestosql/Sequence.cpp:70]: @mbasmanova 
So, I've got a question on how "try" should work.
Say we have 100 rows batch and row 5 has incorrect arguments, so checkArguments() throws.
1. Do we need to only return NULL for that row and process the rest?
2. Do we have a way to stop processing rows if there is no "try" outside and such error happened?

Does Map.cpp have all this?

File: velox/functions/prestosql/Sequence.cpp (hunk 37 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:45]: So without the `try catch` here, some tests are failing on github, and the errors are from ExpressionVerifier "Common eval wasn't supposed to throw, but it did. Aborting."

Then I compared my code with other functions and found they all have the `try catch`. After adding, the tests are passing

File: velox/functions/prestosql/Sequence.cpp [hunk 39]
```diff
@@ -0,0 +1,168 @@
...
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+ private:
+  static VectorPtr applyNumber(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) {
+    const auto numRows = rows.size();
```
Comment (laithsakka) [inline velox/functions/prestosql/Sequence.cpp:58]: uses rows.end()

**Review by laithsakka — COMMENTED** [review]
is there a reason why you choose not to use the simple function interface?

File: velox/functions/prestosql/Sequence.cpp [hunk 40]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  static void checkArguments(int64_t start, int64_t stop, int64_t step) {
+    VELOX_USER_CHECK_NE(step, 0, "step must not be zero");
+    VELOX_USER_CHECK(
+        step > 0 ? stop >= start : stop <= start,
+        "sequence stop value should be greater than or equal to start value if "
+        "step is greater than zero otherwise stop should be less than or equal to start")
+    VELOX_USER_CHECK_LE(
+        (stop - start) / step + 1,
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:104]: nit: to avoid computing this twice, consider returning this value from checkArguments.

File: velox/functions/prestosql/Sequence.cpp [hunk 41]
```diff
@@ -0,0 +1,152 @@
...
+      elements[elementsOffset++] = value;
+    }
+  }
+};
+
+} // namespace
+
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+  signatures.reserve(2);
```
Comment (mbasmanova) [inline velox/functions/prestosql/Sequence.cpp:133]: do this:

```
signatures = {
exec::FunctionSignatureBuilder()
                              .returnType("array(bigint)")
                              .argumentType("bigint")
                              .argumentType("bigint")
                              .build(),
exec::FunctionSignatureBuilder()
                              .returnType("array(bigint)")
                              .argumentType("bigint")
                              .argumentType("bigint")
                              .argumentType("bigint")
                              .build()
};
```

**Review by mbasmanova — COMMENTED** [review]
@kewang1024 Ke, thank you for iterating on this PR. The code looks good, but I haven't gone other the tests yet.

File: velox/functions/prestosql/Sequence.cpp [hunk 42]
```diff
@@ -0,0 +1,152 @@
...
+        "step is greater than zero otherwise stop should be less than or equal to start")
+    VELOX_USER_CHECK_LE(
+        (stop - start) / step + 1,
+        kMaxResultEntries,
+        "result of sequence function must not have more than 10000 entries");
+  }
+
+  static void writeToElements(
+      int64_t* elements,
+      vector_size_t elementsOffset,
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:111]: Wouldn't that move the loop to the outside of the function? It seems to me it would be more compact this way?
